src/v/jl/jls_writer/jls_writer.h

```diff
@@ -0,0 +1,64 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#pragma once
+
+#include "base/likely.h"
+#include "base/seastarx.h"
+#include "bytes/iobuf.h"
+#include "jl/jls/jls.h"
+#include "model/fundamental.h"
+#include "model/record.h"
+
+#include <seastar/core/iostream.hh>
+
+namespace jl::jls_writer {
+inline constexpr auto CHUNK_SIZE_TARGET = 4 * 1024 * 1024;
+
+struct options {
+    jl::jls::chunk_size_t chunk_size_target = CHUNK_SIZE_TARGET;
+};
+
+/// This writer expects data to be already ordered by ntp and offset.
+/// It will throw an exception if the data is not ordered.
+class writer {
+public:
+    explicit writer(ss::output_stream<char>* out, options opts = {})
+      : _out(out)
+      , _opts(opts) {}
+
+public:
+    ss::future<> write(model::ntp ntp, model::record_batch&& batch);
```

Comment:
it'd be more convenient to have low level interface instead, something like
- write_ntp_header(ntp, offset range) (writes per-ntp header to the iobuf)
- write_batches (accepts model::record_batch or record batch reader and copies data)
- write_footer (generates footer and writes it to the iobuf)

Reply:
current implementation accumulates the whole chunk in memory, instead of that every method should just produce to the output stream (we already have the data accumulated in memory at this point so this is a duplication)

Reply:
I guess that we will have a little problem here. We need to know object size in advance in order to upload it to S3. But streaming interface still makes sense in this case. We can just copy the stream into iobuf and then get its size and upload it.